pkg/daemon/update.go

```diff
@@ -1147,42 +1147,57 @@ func (dn *Daemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig) error
 
 	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-extra"}
 	realtimeKernel := []string{"kernel-rt-core", "kernel-rt-modules", "kernel-rt-modules-extra", "kernel-rt-kvm"}
-	var args []string
 
 	dn.logSystem("Initiating switch from kernel %s to %s", canonicalizeKernelType(oldConfig.Spec.KernelType), canonicalizeKernelType(newConfig.Spec.KernelType))
 
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault {
-		args = []string{"override", "reset"}
+	// resetRealTimeOverride removes the realtime kernel override.
+	resetRealTimeOverride := func(msg string) error {
+		args := []string{"override", "reset"}
 		args = append(args, defaultKernel...)
 		for _, pkg := range realtimeKernel {
 			args = append(args, "--uninstall", pkg)
 		}
-		dn.logSystem("Switching to kernelType=%s, invoking rpm-ostree %+q", newConfig.Spec.KernelType, args)
+		dn.logSystem("%s, invoking rpm-ostree %+q", msg, args)
 		_, err := runGetOut("rpm-ostree", args...)
 		return err
 	}
 
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime {
-		// Switch to RT kernel
-		args = []string{"override", "remove"}
+	// setRealTimeOverride installs the realtime kernel and remove the
+	// default kernel.
+	setRealTimeOverride := func(msg string) error {
+		args := []string{"override", "remove"}
 		args = append(args, defaultKernel...)
 		for _, pkg := range realtimeKernel {
 			args = append(args, "--install", pkg)
 		}
-
-		dn.logSystem("Switching to kernelType=%s, invoking rpm-ostree %+q", newConfig.Spec.KernelType, args)
+		dn.logSystem("%s, invoking rpm-ostree %+q", msg, args)
 		_, err := runGetOut("rpm-ostree", args...)
 		return err
 	}
 
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime {
+	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime {
+		// Switch from the realtime to default kernel
+		if canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault {
+			return resetRealTimeOverride(fmt.Sprintf("Switching to kerneltype=%s", newConfig.Spec.KernelType))
+		}
+
+		// Upgrade of realtime kernel requires resetting and then adding back
+		// in the override. rpm-ostree only performs "local" replacements of overrides, which
+		// mean detecting the NVR of the package. In theory, we could just find the path to the
+		// rpm's in the extension repo, however it would bug prone due to dependencies.
 		if oldConfig.Spec.OSImageURL != newConfig.Spec.OSImageURL {
-			dn.logSystem("Updating rt-kernel packages on host: %+q", args)
-			_, err := runGetOut("rpm-ostree", args...)
```

Comment:
This was the root-cause: `rpm-ostree` was called with an empty `[]string`.

Reply:
ah yeah passing empty args here is of no use and caused the bug. Simple way to fix it is just pass `update` as arg to rpm-ostree. Since kernel-rt packages are installed from extensions repo, package dependencies gets automatically handled.

Reply:
Indeed, repushed :)